### PR TITLE
New syscheck fields API

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
|#6398 |

This PR is related to the issue #6398 . I have modified the` GET /syscheck/{agent_id}` endpoint. 

I have added the new fields returned by the syscheck scans to our API responses. I have also added the corresponding filters that will make the API search for syscheck results. The docstrings have been updated to NumpyDoc, too.

The new keys the endpoint will return are `arch`, `value_name` and `value_type`. These keys are generated by Windows registries. In order to replicate this Windows information, I have added examples values to some FIM scans so the API response will have those fields.

- Examples:

These are some examples added to the database in order to test the endpoint. They have been also added to the db when passing the API integration tests and the unit tests. I have taken the next things into account in order to insert the examples to the `fim_entry` table: 

_value_name is NULL when type is 'file' or 'registry_key'
value_type is NULL when type is 'file' or 'registry_key' or the value is invalid
arch is NULL when type is 'file'_

**Files:**

```
INSERT INTO fim_entry VALUES('/sbin/runit','/sbin/runit','file',1578640725,1,NULL,NULL,NULL,18840,'100755','0','0','c80c63cc6759381819691a987f1d7683','a6cca2f27bc8be45d55247cd06d551d15091d226','root','root',1420542776,15470536,'47d90904042bf4557089f08cc287ff5af4048beaec56f37f31ade5d5c3489d25',0,NULL,NULL);
INSERT INTO fim_entry VALUES('/usr/bin/getconf','/usr/bin/getconf','file',1578640719,1,NULL,NULL,NULL,22944,'100755','0','0','7db6b10d542510318ba9a94e369fde26','a3901716d1a678e21398dcd0403a3b36bc1d07cb','root','root',1549397572,16393804,'72f5c4543d4bbe82c80c5d7d4e9c232d4414760aba7c3fc629eb3f91c25599bd',0,NULL,NULL);
INSERT INTO fim_entry VALUES('/usr/bin/frm.mailutils','/usr/bin/frm.mailutils','file',1578640720,1,NULL,NULL,NULL,23296,'100755','0','0','294a5e70b07ba6d0e8a8b6ad59622e94','ae7add46c901ce949d1e80defc5136e6b5d3a644','root','root',1459864318,16393800,'962d7e4d40c5076dc1e2ab2d1c3da539017e857d121cc6a4e6a105044bd10618',0,NULL,NULL);
INSERT INTO fim_entry VALUES('/usr/bin/apt-key','/usr/bin/apt-key','file',1578640718,1,NULL,NULL,NULL,20599,'100755','0','0','50141b833e183a0ea826ed500e25b8f1','f5792d780476ca9033c230937017532cfd1a4461','root','root',1558469542,16393728,'760885b6142d0c6e52df29cffa9026d1cfbc928b9d123979951e3ba547b5937b',0,NULL,NULL);
```

**Registries:**

```
INSERT INTO fim_entry VALUES('registry_key_1','registry_key_1','registry_key',1578640718,1,'[x64]',NULL,NULL,4096,'perm','0','0','','','root','root',12345678,1024,'',0,NULL,NULL);
INSERT INTO fim_entry VALUES('registry_value_1','registry_value_1','registry_value',1578640718,1,'[x32]','value_name','value_type',4096,'perm','0','0','','','root','root',12345678,1024,'',0,NULL,NULL);
```

- API requests examples:

**Request:**

`curl -k -X GET "https://localhost:55000/syscheck/000?type=registry_key&pretty=true" -H "Authorization: Bearer $TOKEN"
`

**Response:**

```
{
  "data": {
    "affected_items": [
      {
        "gid": "0",
        "uname": "root",
        "gname": "root",
        "perm": "perm",
        "sha256": "",
        "arch": "[x64]",
        "mtime": "1970-05-23T21:21:18Z",
        "attributes": 0,
        "size": 4096,
        "uid": "0",
        "sha1": "",
        "date": "2020-01-10T07:18:38Z",
        "type": "registry_key",
        "md5": "",
        "inode": 1024,
        "changes": 1,
        "file": "registry_key_1"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "FIM findings of the agent were returned",
  "error": 0
}
```

**Request:**

`curl -k -X GET "https://localhost:55000/syscheck/000?pretty=true&arch=[x64]" -H "Authorization: Bearer $TOKEN"
`

**Response:**

```
{
  "data": {
    "affected_items": [
      {
        "changes": 1,
        "type": "registry_key",
        "inode": 1024,
        "file": "registry_key_1",
        "uid": "uid",
        "md5": "",
        "gid": "gid",
        "sha256": "",
        "size": 4096,
        "perm": "perm",
        "attributes": "0",
        "date": "2020-01-10T07:18:38Z",
        "sha1": "",
        "uname": "root",
        "gname": "root",
        "mtime": "1970-05-23T21:21:18Z",
        "arch": "[X64]"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "FIM findings of the agent were returned",
  "error": 0
}
```

**Request:**

`curl -k -X GET "https://localhost:55000/syscheck/000?value_name=value_name&pretty=true" -H "Authorization: Bearer $TOKEN"`

**Response:**

```
{
  "data": {
    "affected_items": [
      {
        "gid": "0",
        "uname": "root",
        "gname": "root",
        "perm": "perm",
        "sha256": "",
        "arch": "[x32]",
        "mtime": "1970-05-23T21:21:18Z",
        "attributes": 0,
        "value_type": "value_type",
        "value_name": "value_name",
        "size": 4096,
        "uid": "0",
        "sha1": "",
        "date": "2020-01-10T07:18:38Z",
        "type": "registry_value",
        "md5": "",
        "inode": 1024,
        "changes": 1,
        "file": "registry_value_1"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "FIM findings of the agent were returned",
  "error": 0
}
```

- API integration test results:

```
test_rbac_black_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings

test_rbac_white_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings

test_syscheck_endpoints.tavern.yaml 
	 35 passed, 39 warnings
```

- Unit test results:

```
framework/wazuh/tests/test_syscheck.py ...................                                                                                                 [100%]

======================================================================= 20 passed in 0.21s =======================================================================
```

Regards,
Manuel.

### CODE REVIEW

**Comment by mcarmona99** [comment]
After the base branch update, some fixes were needed:

- Unit test result:

```
framework/wazuh/tests/test_syscheck.py .................     [100%]
================ 20 passed in 0.74s ================
```

- API integration test results:

```
test_rbac_black_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings

test_rbac_white_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings

test_syscheck_endpoints.tavern.yaml 
	 35 passed, 39 warnings
```

Also tried the API requests examples posted on this issue, they are still working properly.

File: api/api/controllers/syscheck_controller.py [hunk 1]
```diff
@@ -41,37 +49,45 @@ async def put_syscheck(request, agents_list='*', pretty=False, wait_for_complete
...
+    md5 : str
+        Filters files with the specified MD5 checksum.
+    sha1 : str
+        Filters files with the specified SHA1 checksum.
+    sha256 : str
+        Filters files with the specified SHA256 checksum.
+    distinct : bool
+        Look for distinct values.
+    q : str
+        Query to filter results by.
```
Comment (davidjiglesias) [inline api/api/controllers/syscheck_controller.py:85]: Missing new params arch, value_name and value_type

File: api/api/controllers/syscheck_controller.py [hunk 2]
```diff
@@ -126,14 +151,21 @@ async def delete_syscheck_agent(request, agent_id='*', pretty=False, wait_for_co
 
 
 async def get_last_scan_agent(request, agent_id, pretty=False, wait_for_complete=False):
-    """
+    """Run a syscheck scan over the agent_ids
```
Comment (davidjiglesias) [inline api/api/controllers/syscheck_controller.py:154]: Wrong description

File: framework/wazuh/syscheck.py [hunk 3]
```diff
@@ -17,10 +17,17 @@
 
 @expose_resources(actions=["syscheck:run"], resources=["agent:id:{agent_list}"])
 def run(agent_list=None):
-    """Run syscheck scan.
+    """Run a syscheck scan for the specified agents.
```
Comment (davidjiglesias) [inline framework/wazuh/syscheck.py:20]:
```suggestion
    """Run a syscheck scan in the specified agents.
```

File: framework/wazuh/syscheck.py [hunk 4]
```diff
@@ -47,10 +54,17 @@ def run(agent_list=None):
 
 @expose_resources(actions=["syscheck:clear"], resources=["agent:id:{agent_list}"])
 def clear(agent_list=None):
-    """Clear the syscheck database for a list of agents.
+    """Clear an agent's syscheck database.
```
Comment (davidjiglesias) [inline framework/wazuh/syscheck.py:57]:
```suggestion
    """Clear the syscheck database of the specified agents.
```

File: framework/wazuh/syscheck.py [hunk 5]
```diff
@@ -79,10 +93,17 @@ def clear(agent_list=None):
 
 @expose_resources(actions=["syscheck:read"], resources=["agent:id:{agent_list}"])
 def last_scan(agent_list):
-    """Gets the last scan of the agent.
+    """Gets the last scan of an agent.
```
Comment (davidjiglesias) [inline framework/wazuh/syscheck.py:96]:
```suggestion
    """Get the last scan of an agent.
```

File: framework/wazuh/syscheck.py [hunk 6]
```diff
@@ -137,25 +158,43 @@ def last_scan(agent_list):
...
-    :param summary: Returns a summary grouping by filename.
-    :param offset: First item to return.
-    :param limit: Maximum number of items to return.
-    :param sort: Sorts the items. Format: {"fields":["field1","field2"],"order":"asc|desc"}.
-    :param search: Looks for items with the specified string.
-    :param select: Select fields to return. Format: ["field1","field2"].
-    :param q: Query to filter by
-    :param distinct: Look for distinct values
-    :return: AffectedItemsWazuhResult.
+    """Return a list of files from the database that match the filters.
```
Comment (davidjiglesias) [inline framework/wazuh/syscheck.py:161]:
```suggestion
    """Return a list of files from the syscheck database of the specified agents.

File: api/api/controllers/syscheck_controller.py [hunk 7]
```diff
@@ -102,13 +119,21 @@ async def get_syscheck_agent(request, agent_id, pretty=False, wait_for_complete=
...
 
 async def delete_syscheck_agent(request, agent_id='*', pretty=False, wait_for_complete=False):
-    """
-    :param pretty: Show results in human-readable format 
-    :type pretty: bool
-    :param wait_for_complete: Disable timeout response 
-    :type wait_for_complete: bool
-    :param agent_id: Agent ID
-    :type agent_id: str
+    """Run a syscheck scan over the agent_ids
```
Comment (davidjiglesias) [inline api/api/controllers/syscheck_controller.py:122]: Wrong description

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Nested values are not being handled properly. Check syscollector (netiface) WazuhDBQuery as an example. Also check spec.yaml for netiface (rx.packets, etc)

value_type -> value : { type: XX}

File: api/api/controllers/syscheck_controller.py (hunk 1 above)
Comment (mcarmona99) [reply api/api/controllers/syscheck_controller.py:85]: This and the next comments are changes missing on the docstrings, I have changed them now

**Comment by mcarmona99** [comment]
Changes to docstrings and how we handle nested values.
After the canges, there are new responses with nested items. Example:

```
{
  "data": {
    "affected_items": [
      {
        "value": {
          "name": "value_name",
          "type": "value_type"
        },
        "uname": "root",
        "file": "registry_value_1",
        "date": "2020-01-10T07:18:38Z",
        "size": 4096,
        "sha256": "",
        "changes": 1,
        "sha1": "",
        "gname": "root",
        "arch": "[x32]",
        "perm": "perm",
        "inode": 1024,
        "md5": "",
        "attributes": 0,
        "gid": "0",
        "mtime": "1970-05-23T21:21:18Z",
        "uid": "0",
        "type": "registry_value"
      },
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "FIM findings of the agent were returned",
  "error": 0
}
```

**TEST RESULTS:**

-Pass API integration tests and unit tests and update them

**Comment by mcarmona99** [comment]
Now we are using `min_select_fields` in order to show always the `file` key. This way we are not getting empty dictionaries when selecting `value.name`, `value.type`, `arch` and/or `attributes`.

EXAMPLE:

- Request:

`https://localhost:55050/syscheck/000?wait_for_complete=true&limit=4&select=arch,value.name`

- Response:

```
{
  "data": {
    "affected_items": [
      {
        "arch": "[x64]",
        "file": "registry_key_1"
      },
      {
        "value": {
          "name": "value_name"
        },
        "arch": "[x32]",
        "file": "registry_value_1"
      },
      {
        "file": "/usr/bin/c89-gcc"
      },
      {
        "file": "/etc/dpkg/origins/debian"
      }
    ],
    "total_affected_items": 1358,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "FIM findings of the agent were returned",
  "error": 0
}
```

TEST RESULTS:

```
framework/wazuh/tests/test_syscheck.py 
===================== 23 passed in 0.19s ======================

test_syscheck_endpoints.tavern.yaml 
	 35 passed, 39 warnings

test_rbac_black_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings

test_rbac_white_syscheck_endpoints.tavern.yaml 
	 4 passed, 6 warnings
```

**Pull request merged**